substrate/client/rpc-spec-v2/src/transaction/transaction_broadcast.rs

```diff
@@ -0,0 +1,217 @@
+
+		let broadcast_transaction_fut = async move {
+			// There is nothing we could do with an extrinsic of invalid format.
+			let Ok(decoded_extrinsic) = TransactionFor::<Pool>::decode(&mut &bytes[..]) else {
+				return
+			};
+
+			// Flag to determine if the we should broadcast the transaction again.
+			let mut is_done = false;
+
+			while !is_done {
+				// Wait for the next block to become available.
+				let Some(mut best_block_hash) = best_block_import_stream.next().await else {
+					return
+				};
+				// We are effectively polling the stream for the last available item at this time.
+				// The `now_or_never` returns `None` if the stream is `Pending`.
+				//
+				// If the stream contains `Hash0x1 Hash0x2 Hash0x3 Hash0x4`, we want only `Hash0x4`.
+				while let Some(next) = best_block_import_stream.next().now_or_never() {
+					let Some(next) = next else {
+						// Nothing to do if the best block stream terminated.
+						return
+					};
+					best_block_hash = next;
+				}
+
+				let submit =
+					pool.submit_and_watch(best_block_hash, TX_SOURCE, decoded_extrinsic.clone());
+
+				// The transaction was not included to the pool, because it is invalid.
+				// However an invalid transaction can become valid at a later time.
+				let Ok(mut stream) = submit.await else { return };
+
+				while let Some(event) = stream.next().await {
+					match event {
+						// The transaction propagation stops when:
+						// - The transaction was included in a finalized block via
+						//   `TransactionStatus::Finalized`.
+						TransactionStatus::Finalized(_) |
+						// - The block in which the transaction was included could not be finalized for
+						//   more than 256 blocks via `TransactionStatus::FinalityTimeout`. This could
+						//   happen when:
+						//   - the finality gadget is lagging behing
+						//   - the finality gadget is not available for the chain
+						TransactionStatus::FinalityTimeout(_) |
+						// - The transaction has been replaced by another transaction with identical tags
+						// (same sender and same account nonce).
+						TransactionStatus::Usurped(_) => {
+							is_done = true;
+							break;
+						},
+
+						// Dropped transaction may enter the pool at a later time, when other
+						// transactions have been finalized and remove from the pool.
+						TransactionStatus::Dropped |
+						// An invalid transaction may become valid at a later time.
+						TransactionStatus::Invalid => {
+							break;
+						},
+
+						// The transaction is still in the pool, the ready or future queue.
+						TransactionStatus::Ready | TransactionStatus::Future |
+						// Transaction has been broadcasted as intended.
+						TransactionStatus::Broadcast(_) |
+						// Transaction has been included in a block, but the block is not finalized yet.
+						TransactionStatus::InBlock(_) |
+						// Transaction has been retracted, but it may be included in a block at a later time.
+						TransactionStatus::Retracted(_) => (),
```

Comment:
How confident are you that if we see a Retracted event, we'll see another event again for the tx in the future? :)

(In Subxt, we also keep waiting for further events after a Retracted, but I'm not 100% sure whether one will always be emitted!)

Reply:
Coolio; well, I'm happy as is but always a little scared that we won't get any future events because the "old" API doesn't make any promises really!